### PR TITLE
Move backup

### PR DESCRIPTION
**Pull request by etcher123**
This is a proposal to add a feature to `node-gpt`.

This PR adds a function that does two things:
- move the backup GPT to another LBA.
- move the end of the last partition accordingly, so that it ends at the beginning of the backup GPT.

The PR is not perfect at this point, but if you're interested in this feature I can improve it according to your feedback.

#### Use-case

The typical use-case for this function: to grow the last partition on a disk.

Let's think about something like the Etcher installer, that is used to install disk images to a USB media. Usually, developers will ship a disk image of a certain size, and users might install this image on a USB media of the same size or bigger.

Sometimes it makes sense to be able to grow the last partition so that it fills the media (for example, the last partition might be the home partition). Which simply translates to:
1. moving the GPT backup to the end
2. growing the last partition to the maximum size

#### Implementation

One decision I made in this implementation is to make the function smart enough to find out if it acts on the primary GPT or the backup GPT. This is done with `var isPrimary = this.backupLBA == 0 || this.currentLBA < this.backupLBA`. It seems safe to me, but maybe you would prefer two explicit functions instead, eg. `moveBackupPrimary` and `moveBackupBackup`.

Another decision I made is to bundle the two actions of moving the GPT and resizing the last partition into one function. I guess it could be separated in two functions, but it would be more complicated to implement.

Maybe there are other things questionable.

#### Testing it

Here's a procedure to test the `grow.js` example quickly (assuming you use a Linux OS).

    # Create an empty file
    $ fallocate -l 1G disk.img
    $ file disk.img
    disk.img: data

    # Create a GPT partition table and one partition
    $ echo -e "label: gpt\n," | /sbin/sfdisk disk.img
    ...
    $ file disk.img
    disk.img: DOS/MBR boot sector; partition 1 : ID=0xee, start-CHS (0x0,0,2), end-CHS (0x3ff,255,63), startsector 1, 2097151 sectors, extended partition table (last)

    # Verify that fdisk is happy
    $ /sbin/fdisk -l disk.img 
    ...    # no error message

    # Now let's grow the image
    $ fallocate -l 2G disk.img
    $ ls -lh disk.img 
    -rw-r--r-- 1 user user 2.0G Jun 16 10:14 disk.img

    # Verify that: 1) fdisk is not happy 2) the partition is still 1G
    $ /sbin/fdisk -l disk.img
    GPT PMBR size mismatch (4194303 != 8388607) will be corrected by write.
    The backup GPT table is not on the end of the device. This problem will be corrected by write.
    ...
    Device     Start     End Sectors  Size Type
    disk.img1   2048 2097118 2095071 1023M Linux filesystem

    # Move the backup GPT and grow the last partition
    $ nodejs ./example/grow.js disk.img

    # Verify that: 1) fdisk is a bit happier 2) the partition is now 2G
    $ /sbin/fdisk -l disk.img
    GPT PMBR size mismatch (2097151 != 4194303) will be corrected by write.
    ...
    Device     Start     End Sectors Size Type
    disk.img1   2048 4194270 4192223   2G Linux filesystem

TLDR: The example `grow.js` does two things:
- move the GPT backup at the end of the device, therefore gets rid of the message "The backup GPT table is not on the end of the device..."
- grow the last partition so that it fills the device

### CODE REVIEW

**Comment by lurch** [comment]
Given that this basically only has a single use-case, I wonder if it'd be worth also adding an even higher-level `moveBackupGPTToEndOfDisk` utility function? :man_shrugging: 

P.S. I don't claim to fully understand the code, but why does it explicitly exclude block devices? Does that mean that in the example Etcher use-case, you'd need to modify the disk-image _before_ writing it to a disk? In reality you'd probably only want to modify the disk and not the original disk-image?

**Comment by etcher123** [comment]
> Given that this basically only has a single use-case, I wonder if it'd be worth also adding an even higher-level moveBackupGPTToEndOfDisk utility function?

That would require that `node-gpt` knows the size of the disk, which it doesn't. I tried to follow the existing logic here, which is that the library strictly deals with the GPT structure, without knowing anything else. The knowledge of where is this GPT structure (a file, a block device?) seems to be out of scope of the `node-gpt` library.

In order to know the size of the disk, we can do `fs.fstatSync(fd)`, but that only works if the disk is in reality a file. For a real block device you need something else, apparently the library `@ronomon/direct-io`. And I don't think we want to add this dependency to `node-gpt`.

> why does it explicitly exclude block devices?

Because to get the size of the block device you need something like [@ronomon/direct-io](https://github.com/ronomon/direct-io#block-device-size-and-sector-size). This is an async library, and I just don't know how to use it. Tried and failed.

Maybe there are other solutions. But for the scope of the example, I just keep it simple.

Mind you: I'm not a javascript or nodejs developer.

> Does that mean that in the example Etcher use-case, you'd need to modify the disk-image before writing it to a disk? In reality you'd probably only want to modify the disk and not the original disk-image?

Indeed you modify the GPT on the disk, after writing the image to a disk. And it works for a block device. I have a fork of the etcher sdk where I added support for that. It works, but the implementation is not great (as I said, I'm not a javascript person so...)

You can see it in this commit: https://github.com/etcher123/etcher-sdk/commit/456e562451d2954e946363126a4b543f111d2f6d

**Comment by lurch** [comment]
> I'm not a javascript or nodejs developer.

Me neither, please take my uninformed comments with a pinch of salt :slightly_smiling_face:

**Comment by jhermsmeier** [comment]
Hi @etcher123!
Thank you for the PR & the work you've put into this!

> Another decision I made is to bundle the two actions of moving the GPT and resizing the last partition into one function. I guess it could be separated in two functions, but it would be more complicated to implement.

While I appreciate the simplicity of combining the two, it's unfortunately not desirable for a few reasons;

- When writing disk images, the secondary GPT should always be relocated to the end of the disk / block device, irrespective of the partitions' sizes – this is due to 1) [the specification](http://www.intel.com/technology/efi/efiagree.htm) saying so (Section 11.2.2.1, EFI Partition Header) and 2) operating systems / drivers checking for a GPT at that location should the first sectors become unreadable (and if you don't relocate it you can get a conflict / incorrect partitions if there is a GPT from the previous formatting there)

- If a partition's allocated space is changed in the partition table without resizing the file system as well (also, a lot of file systems do not support resizing), the file system drivers may be initialized with incorrect data (i.e. if they need to read structures at the end of the partition)

- The operating system or file system drivers might deny any corrections on the above if a file system is not resizable or "too full" to be shrunken down in size, even though the extra space isn't actually used / allocated to the file system in that partition

So I would prefer relocating the GPTs and resizing partitions to be two separate actions.

**Comment by jhermsmeier** [comment]
> Given that this basically only has a single use-case, I wonder if it'd be worth also adding an even higher-level `moveBackupGPTToEndOfDisk` utility function?

Hey @lurch – this would make sense in a higher level module, but this module doesn't do any I/O at all (except for the runnable examples), so wouldn't know where the end of the disk is. 

> Does that mean that in the example Etcher use-case, you'd need to modify the disk-image _before_ writing it to a disk? In reality you'd probably only want to modify the disk and not the original disk-image?

Nah, you can still do any modifications / relocations after – no need to fiddle with the input disk image, it's just that this module is only concerned with parsing, modifying & encoding the GPT related data structures, not doing any I/O – so those choices are up to the program using this, if that makes sense?

File: lib/gpt.js [hunk 1]
```diff
@@ -513,6 +513,33 @@ GPT.prototype = {
...
     return buffer
 
+  },
+
+  /**
+   * Move the backup GPT to another LBA. In the process, also resize
+   * the last partition so that it ends just before the backup GPT.
+   * @param {Number} [newLBA] - this is the LBA of the backup GPT header
+   */
+  moveBackup( newLBA ) {
```
Comment (jhermsmeier) [inline lib/gpt.js:523]: I feel like `relocate( newLBA )` would be a better name (?), since this doesn't just apply to backups / secondaries

File: lib/gpt.js [hunk 2]
```diff
@@ -513,6 +513,33 @@ GPT.prototype = {
...
+      this.tableOffset = newLBA - 32
+    }
+
+    // Set last LBA
+    this.lastLBA = newLBA - 33
+
+    // Set last LBA of the last partition
+    var nparts = this.partitions.length
+    if( nparts > 0 ) {
+      this.partitions[nparts - 1].lastLBA = this.lastLBA
```
Comment (jhermsmeier) [inline lib/gpt.js:541]: As mentioned in a previous comment, I'd prefer if resizing partitions would be a separate method

**Comment by elboulangero** [comment]
Thanks for the feedback. I'll rework that to split in two functions.

> If a partition's allocated space is changed in the partition table without resizing the file system as well (also, a lot of file systems do not support resizing), the file system drivers may be initialized with incorrect data (i.e. if they need to read structures at the end of the partition)

I didn't know about that, actually all I know about is the `ext4` filesystem.

Just to clarify my intent: for Linux OS that boot with systemd, there's an option to automatically grow the filesystem so that it fills the partition: https://www.freedesktop.org/software/systemd/man/systemd.mount.html#x-systemd.growfs

So the scenario where the installer (ie. Etcher) grows the last partition so that it fills the device, WITHOUT resizing the file system, is actually supported and expected by systemd. The filesystem itself will be expanded when the OS is booted for the first time.

In any case, I understand that filesystem support is out of scope here.

Thanks for the feedback!

File: lib/gpt.js (hunk 1 above)
Comment (elboulangero) [reply lib/gpt.js:523]: Agreed

File: lib/gpt.js [hunk 3]
```diff
@@ -513,6 +513,33 @@ GPT.prototype = {
...
+  moveBackup( newLBA ) {
+    // Check whether this is a primary or a backup GPT
+    var isPrimary = this.backupLBA == 0 || this.currentLBA < this.backupLBA
+
+    // Set new LBA of the backup GPT
+    if( isPrimary ) {
+      this.backupLBA = newLBA
+    } else {
+      this.currentLBA = newLBA
+      this.tableOffset = newLBA - 32
```
Comment (lurch) [inline lib/gpt.js:532]: See my comments in #8 and #9 - IMHO you can't just assume this to be 32 blocks, you need to calculate it in a similar way to what `tableSize()` does i.e. `Math.max( align( this.entrySize * this.entries , this.blockSize ), GPT.TABLE_MIN_SIZE ) / this.blockSize`

**Comment by etcher123** [comment]
@jhermsmeier Please have a look at the updated MR when you have a moment and tell me what you think, thanks!

File: example/grow.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+var backupGPT = utils.readBackupGPT( fd, primaryGPT )
+
+console.log( 'Backup GPT:', inspect( backupGPT ) )
+console.log( '' )
+
+// Get the size of the raw image
+var stats = fs.fstatSync(fd)
+if( stats.isBlockDevice() ) {
+  console.error( 'Block devices are not supported by this example' )
```
Comment (lurch) [inline example/grow.js:61]: It says here that block devices are not supported by the example, yet the usage-message on line 15 shows it being used with a block device? :wink:

File: example/grow.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+// Relocate the backup GPT
+primaryGPT.relocate( sizeInBlocks - 1 )
+backupGPT.relocate( sizeInBlocks - 1 )
+
+// Grow the last partition then
+primaryGPT.growLastPartition( )
+backupGPT.growLastPartition( )
+
+utils.writePrimaryGPT( fd, primaryGPT )
+utils.writeBackupGPT( fd, backupGPT )
```
Comment (lurch) [inline example/grow.js:82]: Maybe it's also worth adding a `utils.writeBothGPTs` function? Since theoretically they should both be kept in sync on-disk? :slightly_smiling_face: :shrug:

File: lib/gpt.js [hunk 6]
```diff
@@ -518,6 +526,56 @@ GPT.prototype = {
...
+   */
+  relocate( newLBA ) {
+    // Check whether this is a primary or a backup GPT
+    var isPrimary = this.backupLBA == 0 || this.currentLBA < this.backupLBA
+
+    // Set the new LBA of the backup GPT
+    if( isPrimary ) {
+
+      if( newLBA <= this.backupLBA ) {
+        return
```
Comment (lurch) [inline lib/gpt.js:548]: I wonder if this should raise an error, rather than just silently returning? :shrug:

File: example/grow.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+if( sizeInBytes % blockSize != 0 ) {
+  console.log( 'sizeInBytes is not a multiple of blockSize!' )
+  process.exit( 1 )
+}
+var sizeInBlocks = sizeInBytes / blockSize
+
+console.log('File size:', sizeInBytes, 'bytes,', sizeInBlocks, 'sectors')
+
+// Relocate the backup GPT
+primaryGPT.relocate( sizeInBlocks - 1 )
```
Comment (lurch) [inline example/grow.js:74]: `primaryGPT.relocate` implies that you're relocating the primaryGPT? Maybe it makes more sense to rename the method to `relocateBackup` ?

And from a purely-semantic point of view, maybe it makes sense to do `backupGPT.relocateBackup` _before_ `primaryGPT.relocateBackup` ?

File: lib/gpt.js [hunk 8]
```diff
@@ -518,6 +526,56 @@ GPT.prototype = {
...
+   *
+   * According to the UEFI spec, the backup GPT should always be located
+   * at the end of the disk. For us, it means that we only support moving
+   * it after its current location.
+   *
+   * @param {Number} [newLBA] - this is the LBA of the backup GPT header
+   */
+  relocate( newLBA ) {
+    // Check whether this is a primary or a backup GPT
+    var isPrimary = this.backupLBA == 0 || this.currentLBA < this.backupLBA
```
Comment (lurch) [inline lib/gpt.js:542]: I wonder if it might be worth adding this as a property itself to the GPT class? @jhermsmeier ?

File: example/grow.js (hunk 4 above)
Comment (etcher123) [reply example/grow.js:61]: @lurch That's a good catch, I'll update it :)

BTW, if you know how to get the size of a block device, synchronously, then we could support block device in the example. That's the only blocker here. Doing `fs.fstatSync(fd)` and then `stats.size` does not work on block devices. The library [ronomon/direct-io](https://github.com/ronomon/direct-io) provides a [getBlockDevice()](https://github.com/ronomon/direct-io#block-device-size-and-sector-size) that you can use to get the block device size, however it's an asynchronous method, I don't know how to use it...

File: example/grow.js (hunk 7 above)
Comment (etcher123) [reply example/grow.js:74]: Not sure about the right name, I'll let @jhermsmeier decide on that.

Some functions in the library have the `Backup` suffix to indicate if they should be called on a primary GPT or on a backup GPT (for example, `parse()` and `parseBackup()`). So calling the function `relocateBackup()` can also be confusing in its own way.

File: example/grow.js (hunk 5 above)
Comment (etcher123) [reply example/grow.js:82]: In the context of an example, having two lines like this makes it a bit more explicit, in my opinion. But that's a nitpick, doesn't matter to me.

File: example/grow.js (hunk 4 above)
Comment (lurch) [reply example/grow.js:61]: I probably know even less NodeJS than you :wink:

Comment (etcher123) [reply example/grow.js:61]: Updated

Comment (jhermsmeier) [reply example/grow.js:61]: Node's fs API unfortunately doesn't give us a way to determine the block size of a device reliably; as you mentioned there's `direct-io`, which I'd normally also use, but to keep the examples simpler, I've tried to avoid more dependencies.

Since it's stated in the usage text that this assumes a block size of 512 bytes, I think it's "safe" to operate on block devices here.

Btw, `stats.isBlockDevice()` does not necessarily work to filter block devices all the time – depending on platform and device path, it may return `false`, while `stats.isCharacterDevice()` then returns `true` (I don't know the reasons behind that though, never dug into it).

File: example/grow.js (hunk 5 above)
Comment (jhermsmeier) [reply example/grow.js:82]: I'd prefer being explicit, not a fan of layering on needless abstractions

File: lib/gpt.js (hunk 8 above)
Comment (jhermsmeier) [reply lib/gpt.js:542]: Hmm, yes, might be nice to have `.isPrimary` and `.isSecondary` getters, but adding methods requires some more thought and testing – should be done outside of this PR

File: lib/gpt.js (hunk 6 above)
Comment (jhermsmeier) [reply lib/gpt.js:548]: Throwing an error would be a bit much here I think; but returning the offset in bytes it was relocated to would be good, so that one can check and assert on that return value @etcher123

Comment (jhermsmeier) [reply lib/gpt.js:548]: Not sure this behavior is as helpful as it might seem at first glance – i.e. if you want to restore a primary GPT from a secondary, you'd need to relocate the secondary forwards, to the start of the device, and this would prevent doing that easily.

File: lib/gpt.js [hunk 9]
```diff
@@ -518,6 +526,56 @@ GPT.prototype = {
...
+    this.lastLBA = newLBA - this.tableSizeBlocks - 1
+  },
+
+  /**
+   * Grow the last partition (so that it ends at the backup GPT)
+   *
+   * Use with caution. Some filesystems might rely on data written at
+   * the end of the partition, hence resizing it could cause problems.
+   */
+  growLastPartition( ) {
```
Comment (jhermsmeier) [inline lib/gpt.js:571]: Could this be generalized to work for any partition? This method feels a bit too specific to a use case imo. How about `.resizePartition( partitionIndex, firstLBA, lastLBA )`?
That could check for overlapping onto neighbouring partitions, and throw an error in that case.

File: example/grow.js (hunk 7 above)
Comment (jhermsmeier) [reply example/grow.js:74]: The variable naming might be a bit confusing here with relocating the primary to the end of the device, but having `.relocate()` and `.relocateBackup()` makes no sense to me, since they'd do the same thing.

File: lib/gpt.js (hunk 6 above)
Comment (lurch) [reply lib/gpt.js:548]: Feels to me like a hypothetical `createPrimaryFromSecondary` is actually something different to `relocateSecondary` ? :shrug: 

I agree that an error might be a bit much - I just meant that the caller should have some way of knowing whether this function succeeded or not, rather than "silently" doing nothing. And returning the new offset would do that :+1:

File: example/grow.js (hunk 7 above)
Comment (lurch) [reply example/grow.js:74]: I was only suggesting renaming the function, not having two functions with different names, but meh :wink:

File: lib/gpt.js (hunk 9 above)
Comment (lurch) [reply lib/gpt.js:571]: That might be more generalized, but wouldn't changing the firstLBA of a partition always break whatever filesystem is inside that partition? Whereas changing the lastLBA _does_ actually have a practical usage.

I guess this comes down to a purity-vs-pragmatism argument though, which I'll leave up to @jhermsmeier to decide :slightly_smiling_face:

File: example/grow.js (hunk 7 above)
Comment (jhermsmeier) [reply example/grow.js:74]: Even then, relocating (aka just changing the location) is completely decoupled from which one it is or where it previously was. It would only get meaning through the context in which it is done (i.e. relocating a backup to EOD), so the method name should not attach semantics to something where there are none to begin with, if that makes sense

File: lib/gpt.js (hunk 9 above)
Comment (jhermsmeier) [reply lib/gpt.js:571]: > but wouldn't changing the firstLBA of a partition always break whatever filesystem is inside that partition?

That is up to the user of this module to deal with – this deals with the GPT, and the GPT only.
Also, no, it wouldn't always break the file system – i.e. when moving the entire partition back or forth by x blocks the file system should continue to work normally (given it's a file system using relative addresses within itself).

> Whereas changing the lastLBA _does_ actually have a practical usage.

When you're formatting a device you need to put down the partitioning scheme first, before initializing any file systems, so setting the firstLBA is definitely something you'd do during that. Also, you might want to partition a device without sticking any file systems into it at that time

File: example/grow.js (hunk 7 above)
Comment (lurch) [reply example/grow.js:74]: I suppose in that case the `if( newLBA <= this.currentLBA )` check inside `relocate` ought to be removed too?
(i.e. the code @etcher123 originally added _only_ did "relocating a backup to EOD" but @jhermsmeier is requesting it be more generic "just changing the location")

Comment (jhermsmeier) [reply example/grow.js:74]: See https://github.com/jhermsmeier/node-gpt/pull/5#discussion_r444714068

File: example/grow.js (hunk 4 above)
Comment (etcher123) [reply example/grow.js:61]: It's not only a the block size, it's also the size of the block device. It's what I'm talking about here. We need to know the size of the devie if we want to move the backup GPT to the end of it.

File: lib/gpt.js (hunk 6 above)
Comment (etcher123) [reply lib/gpt.js:548]: Just as a reminder, the original purpose of this PR was to allow relocating the secondary GPT to the end of the disk, a specific use-case tailored for installers (like Etcher) that install an image of fixed size on a disk of unknown size (equal or bigger than the image). So the `relocate()` function proposed here did just that originally.

@jhermsmeier I understand you want to make it more generic. But how much?

We can make it "relocate the secondary GPT anywhere with no check", so basically I just remove the checks `newLBA <= this.backupLBA` and `newLBA <= this.currentLBA` and that's it. I'm fine with that, but is there any use-case for that? (Note that, even without obvious use-case, I'm still fine removing the checks, and assume user takes care of not shooting himself in the foot).

I'm not sure to understand the other use-case that you mention though:

> f you want to restore a primary GPT from a secondary, you'd need to relocate the secondary forwards, to the start of the device, and this would prevent doing that easily.

If you'd want to do that, the way I understand it, it would requires two steps:
- `createPrimaryFromSecondary` as @lurch suggests (or maybe a better name would be `writePrimaryFromBackup` to match what already exists in the code). That basically would give you a primary GPT.
- Then just write it at the beginning of the disk? And that's all? There would be nothing else to do.

File: example/grow.js (hunk 7 above)
Comment (etcher123) [reply example/grow.js:74]: @jhermsmeier You're wrong on this one, or maybe I don't understand you:

> Even then, relocating (aka just changing the location) is completely decoupled from which one it is

Just to clarify: the function can be called on both primary or secondary GPT objects, but what it actually relocates, in any case, is the secondary GPT.

It does not relocate the primary GPT. And I think it makes no sense to relocate the primary GPT anyway (it's not like the beginning of the disk can move, or like an image can be installed elsewhere than at the beginning of a disk).

Regarding the name of the function again, if it's clear for everyone that relocating the primary GPT is a nonsense, then naming the function `relocate()` only is kind of OK. It's implicit that we're talking relocating the backup. But one can also prefer to be more explicit and to name it `relocateBackup()` or `relocateSecondary()`. I tend to agree with @lurch on that, now that I think again about it.

File: lib/gpt.js (hunk 8 above)
Comment (etcher123) [reply lib/gpt.js:542]: TBH I'm not sure if this test is robust enough to be made public. It works for me, it makes sense from what I understand, but I wouldn't claim it's enough in all and every situation.

File: lib/gpt.js (hunk 6 above)
Comment (etcher123) [reply lib/gpt.js]: > Throwing an error would be a bit much here I think; but returning the offset in bytes it was relocated to would be good, so that one can check and assert on that return value @etcher123

@jhermsmeier I just did that, see https://github.com/jhermsmeier/node-gpt/compare/c04d81a9ec338edf3460cd4192de52ed4747a73d..9f9d0cb9444dd4194f0d2ee0ced83a50b1b78563

File: lib/gpt.js (hunk 9 above)
Comment (etcher123) [reply lib/gpt.js]: OK I just pushed an update: https://github.com/jhermsmeier/node-gpt/compare/7a429f41ed66017c8a90f3564b7d503801415d08..504c5319f00fb2e23d5ee59dc975100129415f5d

@jhermsmeier If you want I can create a separate PR fr the topic of resizing a partition. The implementation is straightforward enough, however I think we need to add test-cases.

Regarding the implementation, I decided to return the new size in bytes, but maybe you have other preferences (like returning the difference between old size and new size maybe).

Another idea: we could support `-1` for `newLBA` and `lastLBA`, with the meaning `-1 = don't modify it`.

In the example `grow.js`, you can see that I access `primaryGPT.partitions....`, I don't know if it's supposed to be done this way, whether fields are supposed to be public or private. JS doesn't yell at me so apparently it's OK.

**Comment by 1Shola1** [comment]
can i join this project please?
